File: calfflac/calfflac.cpp
/*
ID: cussyou1
PROG: calfflac
LANG: C++
*/
#include <iostream>
#include <fstream>

using namespace std;

int const MaxOfStr=20000;
struct node{
    int start;
    int end;
};
// ignore case
bool compNode( char* array, node a, node b)
{
    if((a.end-a.start) != (b.end-b.start)){
        return false;
    }
    int len = a.end-a.start;
    for(int i=0;i<len;i++){
        if(array[a.start+i] != array[b.start+i]) return false;
    }
    return true;
}
int main()
{
    // all varibles
    char input[MaxOfStr];
    int inputSize;
    // save the real charector only a-z A-Z
    char charArray[MaxOfStr];
    // save the position of every char in the orignal input string
    int charPosition[MaxOfStr];
    int arraySize;
    node palnArray[MaxOfStr];
    int palnSize;

    // read all
    ifstream fin("calfflac.in");
    inputSize =0;
    while(!fin.eof()){
        fin>>input[inputSize];
        ++inputSize;
    }
    // transform
    arraySize=0;
    for(int i=0;i<inputSize;++i){
        if(((input[i]>='a')&&(input[i]<='z'))||((input[i]>='A')&&(input[i]<='Z'))){
            if((input[i]>='A')&&(input[i]<='Z')){
                charArray[arraySize]=input[i]-'A'+'a';
            }else{
                charArray[arraySize]=input[i];
            }
            charPosition[arraySize]=i;
            ++arraySize;
        }
    }
    // ok, let's go

    palnSize=0;
    int k=0;
    bool unit = false; // whether unit 2 or 3 node into 1
    while(k<arraySize||unit){
        unit = false;
        if(palnSize>2){
            bool ret = compNode(charArray,palnArray[palnSize-3],palnArray[palnSize-1]);
            if(ret){
                unit=true;
                palnArray[palnSize-3].end=palnArray[palnSize-1].end;
                palnSize=palnSize-2;
            }else{
                unit=false;
            }
        }
        if(palnSize>1){
             bool ret = compNode(charArray,palnArray[palnSize-2],palnArray[palnSize-1]);
             if(ret){
                 unit=true;
                 palnArray[palnSize-2].end=palnArray[palnSize-1].end;
                palnSize=palnSize-1;
             }else{
                 unit=false;
             }
        }
        if(!unit){ // forward a new node
            palnArray[palnSize].end=k+1;
            palnArray[palnSize].start=k;
            ++k;
            ++palnSize;
        }
    }
    int max=0;
    int maxIndx=0;
    for(int i=0;i<palnSize;++i){
        int tmpMax=palnArray[i].end-palnArray[i].start;
        if(tmpMax>max){
            max=tmpMax;
            maxIndx=i;
        }
    }
    ofstream fout("calfflac.out");
    fout<<max<<endl;
    for(int i=charPosition[palnArray[maxIndx].start];i<charPosition[palnArray[maxIndx].end];++i){
        fout<<input[i];
    }
    fout<<endl;
    return 0;
}

File: calfflac/calfflac.pas
{
}
program calfflac (input,output);
{$APPTYPE CONSOLE}

type 
	palnNode = record
		sindx: integer;
		eindx: integer; 
	end;

var
	{ the orignal text }
	textArray : array [1..20000] of char;
	textSize : integer;
	{ the text only a-z A-Z }
	charArray : array [1..20000] of char;
	charPstn : array [1..20000] of integer;
	charSize : integer;
	tmp : char;
	tmpNode : palnNode;
	palnArray : array[1..20000] of palnNode;
	palnSize : integer;
	i : integer;
	max : integer;
	maxIndx : integer;
function compPaln( a: palnNode; b: palnNode) :boolean;
var 
	i: integer;
	d: integer;
begin
	compPaln:=true;
	if a.eindx-a.sindx <> b.eindx-b.sindx then
		compPaln:=false
	else
		begin
			d:= b.sindx-a.sindx;
			for i:=a.sindx to a.eindx do
				if charArray[i] <> charArray[i+d] then
					begin
						compPaln:=false;
						break;
					end;
		end;
end;
begin
	assign(input,'calfflac.in');
	assign(output,'calfflac.out');
	reset(input);
	rewrite(output);
	{ read }
	textSize:=0;
	charSize:=0;
	while not eof(input) do
	begin
		read(tmp);
		textArray[textSize]:=tmp;
		if ((tmp>='a')and(tmp<='z')) or ((tmp>='A')and(tmp<='Z')) then
			begin
				charArray[charSize+1]:=tmp;
				charPstn[charSize+1]:=textSize;{ save the position of this char }
				inc(charSize);
			end;
		inc(textSize);
	end;
	{ init the palnArray }
	for i:=1 to charSize do 
		begin
			palnArray[i].sindx:=0;
			palnArray[i].eindx:=0;
		end;
	{ ok let's go }
	palnSize:=1;
	tmpNode.sindx:=1;
	tmpNode.eindx:=1;
	palnArray[1]:=tmpNode;
	for i:=2 to charSize do
		begin
			tmpNode.sindx:=i;
			tmpNode.eindx:=i;
			if compPaln(tmpNode,palnArray[palnSize]) then { find a palindrom }
				palnArray[palnSize].eindx:=tmpNode.eindx { add to one palindrom }
			else if palnSize>2 then 
				if compPaln(tmpNode,palnArray[palnSize-1]) then { find a palindrom }
				begin
					dec(palnSize);
					palnArray[palnSize].eindx:=tmpNode.eindx;
				end
			else
				inc(palnSize);
		end;	
	{ find max }
	max:=0;
	for i:=1 to palnSize do
		if palnArray[i].eindx-palnArray[i].sindx > max then 
		begin
			max:= palnArray[i].eindx-palnArray[i].sindx ;
			maxIndx := i
		end;
	writeln(max);
	for i:=charPstn[palnArray[maxIndx].sindx] to charPstn[palnArray[maxIndx].eindx] do 
		write(textArray[i]);
	writeln;
	close(input);
	close(output);
end.
